### PR TITLE
TAO-7903 audio diagnostic

### PR DESCRIPTION
**Pull request by rkarolis**
https://oat-sa.atlassian.net/browse/TAO-7903

Dependancy:
https://github.com/oat-sa/extension-tao-depp/pull/133

Add this configuration in `config/tao/client_lib_config_registry.conf.php` at the end :
```
'taoClientDiagnostic/component/diagnostic/diagnosticLoader' => array(
            'diagnostics' => array(
                'audioDiagnostic' => array(
                    'id' => 'audiorecording',
                    'module' => 'taoDepp/tools/diagnostic/audioDiagnostic',
                    'bundle' => 'taoDepp/loader/diagnostic.min',
                    'position' => null,
                    'name' => 'audioDiagnostic',
                    'description' => 'audio recording diagnostic',
                    'category' => 'diagnostic',
                    'active' => true,
                    'tags' => array()
                ),
                'diagnostic' => array(
                    'id' => 'default',
                    'module' => 'taoClientDiagnostic/tools/diagnostic/diagnostic',
                    'bundle' => 'taoClientDiagnostic/loader/diagnostic.min',
                    'position' => null,
                    'name' => 'diagnostic',
                    'description' => 'default diagnostic',
                    'category' => 'diagnostic',
                    'active' => true,
                    'tags' => array()
                )
            )
        )
```

Some changes should be done in `config/taoClientDiagnostic/clientDiag.conf.php` in order to add new tester and group them, so that relavant testers would be loaded for each factory.

### CODE REVIEW

File: views/scss/diagnostics.scss [hunk 1]
```diff
@@ -13,120 +13,130 @@
...
 
-        label {
-            display: inline-block;
-            width: 150px;
+            label {
+                display: inline-block;
+                width: 150px;
+            }
+        }
+        .hearing-results {
```
Comment (jsconan) [inline views/scss/diagnostics.scss]: Should not this be specific to the other PR? seems related the audio stuff.

**Comment by jsconan** [comment]
The existing diagnostic does not seem to work properly as it hangs on the performances test. Then in the network panel I can see this request `http://ce.tao/taoClientDiagnostic/DiagnosticChecker/storeData` receiving a 403 status.

In `develop` all is working fine. So I'm pretty sure this is related to this PR

File: scripts/update/Updater.php [hunk 2]
```diff
@@ -728,5 +728,43 @@ public function update($initialVersion)
...
+            ];
+            $extension->setConfig('client_lib_config_registry', $clientLibConfig);
+
+            // Update clientDiag.conf.php
+            $extension = $this->getServiceManager()
+                ->get(\common_ext_ExtensionsManager::SERVICE_ID)
+                ->getExtensionById('taoClientDiagnostic');
+            $oldClientDiagConfig = $extension->getConfig('clientDiag');
+            // move value of diagHeader key under header
+            $diagnostic['header'] = $oldClientDiagConfig['diagHeader'];
```
[the remark this answers is not in the document]
Comment (gyszucs) [reply scripts/update/Updater.php]: What you mean exactly?

File: scripts/update/Updater.php [hunk 3]
```diff
@@ -728,5 +728,43 @@ public function update($initialVersion)
...
         $this->skip('2.17.9', '4.1.0');
+
+        if ($this->isVersion('4.1.0')) {
+
+            // Update client_lib_config_registry.conf.php
+            $extension = $this->getServiceManager()
+                ->get(\common_ext_ExtensionsManager::SERVICE_ID)
+                ->getExtensionById('tao');
+            $clientLibConfig = $extension->getConfig('client_lib_config_registry');
+            $clientLibConfig['taoClientDiagnostic/component/diagnostic/diagnosticLoader']['diagnostics']['taoClientDiagnostic/tools/diagnostic/diagnostic'] = [
```
[the remark this answers is not in the document]
Comment (gyszucs) [reply scripts/update/Updater.php]: I wasn't aware of this. Seems it is the right solution. Done

File: scripts/update/Updater.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsconan) [reply scripts/update/Updater.php]: The variable `$diagnostic` comes from the outer space. Currently it is not in use elsewhere, so you don't have issue right now. However, it is unsafe to use derreferenced variable without proper initialization.

[the remark this answers is not in the document]
Comment (gyszucs) [reply scripts/update/Updater.php]: done

File: scripts/update/Updater.php [hunk 4]
```diff
@@ -728,5 +728,43 @@ public function update($initialVersion)
...
+            $extension = $this->getServiceManager()
+                ->get(\common_ext_ExtensionsManager::SERVICE_ID)
+                ->getExtensionById('taoClientDiagnostic');
+            $oldClientDiagConfig = $extension->getConfig('clientDiag');
+            // move value of diagHeader key under header
+            $diagnostic['header'] = $oldClientDiagConfig['diagHeader'];
+            unset($oldClientDiagConfig['diagHeader']);
+            $newClientDiagConfig = [
+                'diagnostic' => array_merge($diagnostic, $oldClientDiagConfig),
+                'audioDiagnostic' => []
```
[the remark this answers is not in the document]
Comment (gyszucs) [reply scripts/update/Updater.php]: fixed

File: manifest.php [hunk 5]
```diff
@@ -29,7 +29,7 @@
     'label'       => 'Browser and OS diagnostic tool',
     'description' => 'Check compatibility of the os and browser of a client',
     'license'     => 'GPL-2.0',
-    'version'     => '4.1.0',
+    'version'     => '4.2.0',
```
[the remark this answers is not in the document]
Comment (gyszucs) [reply manifest.php]: Good catch. Fixed.

File: controller/CompatibilityChecker.php [hunk 6]
```diff
@@ -47,15 +47,10 @@ public function index()
...
-
             $config = $this->loadConfig();
-            if (isset($config['diagHeader'])) {
-                $config['header'] = $config['diagHeader'];
-                unset($config['diagHeader']);
-            }
 
-            if (!empty($config['pageTitle'])) {
-                $this->setData('title', $config['pageTitle']);
+            if (!empty($config['diagnostic']['pageTitle'])) {
```
[the remark this answers is not in the document]
Comment (gyszucs) [reply controller/CompatibilityChecker.php]: No worries, it will return false in such cases.

File: views/js/component/diagnostic/tpl/component.tpl [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<div class="diagnostics-content-area">
+    <h1>Diagnostic Tool</h1>
```
Comment (jsconan) [inline views/js/component/diagnostic/tpl/component.tpl]: This should be translated: `{{__ 'Diagnostic Tool'}}`. This is why the translation is lost in the `*.po` files.
Don't forget to reapply an update of translations after.

File: views/js/tools/diagnostic/diagnostic.js [hunk 8]
```diff
@@ -484,15 +485,20 @@ define([
                 }
             });
         }
+        this.config = getConfig(config, _defaults);
```
Comment (jsconan) [inline views/js/tools/diagnostic/diagnostic.js]: This line is incorrect. The `this` keyword cannot be used here, we are in a factory, not in an object.
Then this is not the place to init the config. The default values are set to the `component` factory as this is the second parameter: `component(specs, defaults)`.

Then the config must be passed to the `init()` method.

File: views/js/tools/diagnostic/diagnostic.js [hunk 9]
```diff
@@ -728,9 +734,20 @@ define([
                     $btn.addClass('hidden');
                     $result.find('[data-action="show-details"]').removeClass('hidden');
                 });
-            })
-            .init(config);
+            });
+
+        _.defer(function() {
+            diagComponent.init();
```
Comment (jsconan) [inline views/js/tools/diagnostic/diagnostic.js]: The `config` must be provided here: `diagComponent.init(config);`

File: scripts/install/RegisterDiagnosticComponent.php [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2018 (original work) Open Assessment Technologies SA;
```
Comment (jsconan) [inline scripts/install/RegisterDiagnosticComponent.php]: Wrong date?

File: views/js/controller/Diagnostic/diagnostic.js [hunk 11]
```diff
@@ -55,7 +55,7 @@ define([
             var extension = $container.data('extension') || 'taoClientDiagnostic';
             var $list = $container.find('.list');
             var $panel = $('.panel');
-            var config = $container.data('config');
+            var config = $container.data('config')['diagnostic'];
```
Comment (jsconan) [inline views/js/controller/Diagnostic/diagnostic.js]: could not it be managed by the backend?

File: views/js/controller/Diagnostic/index.js [hunk 12]
```diff
@@ -118,7 +118,7 @@ define([
             var extension = $container.data('extension') || 'taoClientDiagnostic';
             var $list = $container.find('.list');
             var dataset = $container.data('set');
-            var config = $container.data('config') || {};
+            var config = $container.data('config')['diagnostic'] || {};
```
Comment (jsconan) [inline views/js/controller/Diagnostic/index.js]: Could not it be managed by the backend instead?

File: scripts/install/RegisterDiagnosticComponent.php [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * @param $params
+     * @return \common_report_Report
+     */
+    public function __invoke($params)
+    {
+        ClientLibConfigRegistry::getRegistry()
+            ->register('taoClientDiagnostic/component/diagnostic/diagnosticLoader', [
+                'diagnostics' => [
+                    'taoClientDiagnostic/tools/diagnostic/diagnostic' => [
```
Comment (jsconan) [inline scripts/install/RegisterDiagnosticComponent.php]: Sorry for the late change, but I guess this could be simplified. 
This full namespace is not required here, and it could be replaced by the same key as in the config. That will simplify the architecture.

```php
'diagnostics' => [
    'diagnostic' => [
        'id' => 'default',
        'module' => 'taoClientDiagnostic/tools/diagnostic/diagnostic',
        'bundle' => 'taoClientDiagnostic/loader/diagnostic.min',
        'position' => null,
        'name' => 'diagnostic',
        'description' => 'default diagnostic',
        'category' => 'diagnostic',
        'active' => true,
        'tags' => []
    ]
]
```

File: scripts/update/Updater.php [hunk 14]
```diff
@@ -728,5 +729,39 @@ public function update($initialVersion)
         }
 
         $this->skip('2.17.9', '4.1.0');
+
+        if ($this->isVersion('4.1.0')) {
+            // Update client_lib_config_registry.conf.php
+            ClientLibConfigRegistry::getRegistry()
+                ->register('taoClientDiagnostic/component/diagnostic/diagnosticLoader', [
+                    'diagnostics' => [
+                        'taoClientDiagnostic/tools/diagnostic/diagnostic' => [
```
Comment (jsconan) [inline scripts/update/Updater.php]: Sorry for the late change, but it should be better to only use the config key here: `diagnostic`.

File: views/js/tools/diagnostic/diagnostic.js [hunk 15]
```diff
@@ -728,9 +734,20 @@ define([
...
+        return diagComponent;
     }
 
-    return diagnosticFactory;
+    /**
+     * Wraps the diagnostic factory to be loaded in diagnostic component as a module
+     */
+    return {
+        name: 'diagnostic',
+        init: diagnosticFactory
```
Comment (jsconan) [inline views/js/tools/diagnostic/diagnostic.js]: This could be simplified. The name should not be managed here. That introduces complexity and could lead to mapping issues. Moreover, that also introduces unnecessary changes on existing components.

File: views/js/controller/Diagnostic/diagnostic.js [hunk 16]
```diff
@@ -68,7 +68,7 @@ define([
              * @param {String} [workstation]
              */
             function installTester(workstation) {
-                diagnosticFactory(config)
+                diagnosticFactory.init($list, config)
```
Comment (jsconan) [inline views/js/controller/Diagnostic/diagnostic.js]: The use of `.init` could be avoided

File: views/js/component/diagnostic/diagnosticLoader.js [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+                moduleLoader()
+                    .addList(moduleConfig.diagnostics)
+                    .load(context.bundle)
+                    .then(function(factories) {
+                        _.forEach(factories, function (factory) {
+                            var componentConfig = self.getConfig();
+                            var factoryConfig =  componentConfig[factory.name];
+                            factoryConfig.controller = componentConfig.controller;
+
+                            factory.init(self.getElement(), factoryConfig);
```
Comment (jsconan) [inline views/js/component/diagnostic/diagnosticLoader.js]: The use of `.init` can be avoided.
The wrapping of the factories into an object to add the name introduces unnecessary complexity.
It would be better to use the config keys instead. On the backend the key for each module to load does not need to use the full URI, so it could be the same key as for the config.
Then, this loader could be updated to rely on the keys as module name:
```javascript
return function diagnosticLoaderFactory(container, config) {
        var api = {};
        var component = componentFactory(api, defaults)
        // set the component's layout
            .setTemplate(componentTpl)

        // auto render on init
            .on('init', function(){
                this.render(container);
            })
            // renders the component
            .on('render', function () {
                var self = this;
                var moduleConfig = module.config();
                var identifiers = _.keys(moduleConfig.diagnostics);
                /*
                 * This loads all the modules from module configuration, which are in the `diagnostics` array.
                 */
                moduleLoader({}, _.isFunction)
                    .addList(moduleConfig.diagnostics)
                    .load(context.bundle)
                    .then(function(factories) {
                        _.forEach(factories, function (factory, index) {
                            var componentConfig = self.getConfig();
                            var factoryName = identifiers[index];
                            var factoryConfig =  componentConfig[factoryName];
                            factoryConfig.controller = componentConfig.controller;

                            factory(self.getElement(), factoryConfig);
                        });

                        /**
                         * @event ready
                         */
                        self.trigger('ready');
                    });
            });

        _.defer(function() {
            component.init(config);
        });

        return component;
    };
```

File: views/js/component/diagnostic/diagnosticLoader.js [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+                 * This loads all the modules from module configuration, which are in the `diagnostics` array.
+                 */
+                moduleLoader()
+                    .addList(moduleConfig.diagnostics)
+                    .load(context.bundle)
+                    .then(function(factories) {
+                        _.forEach(factories, function (factory) {
+                            var componentConfig = self.getConfig();
+                            var factoryConfig =  componentConfig[factory.name];
+                            factoryConfig.controller = componentConfig.controller;
```
[the remark this answers is not in the document]
Comment (rkarolis) [reply views/js/component/diagnostic/diagnosticLoader.js]: This is done because it needs to know the controller of the .php page we are in, otherwise API request location will be incorrect.

File: views/js/controller/Diagnostic/index.js (hunk 12 above)
Comment (gyszucs) [reply views/js/controller/Diagnostic/index.js]: No idea how at the moment.

File: views/js/component/diagnostic/diagnosticLoader.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (jsconan) [reply views/js/component/diagnostic/diagnosticLoader.js]: Ok, but that seems weird to not be able to keep the controller set for the module. Anyway, we should keep that in mind if that leads to issue later.
